### PR TITLE
Particle helper

### LINKED ISSUE #2445: Create a particles helper library
Something like createFire or createSmoke

### PR DESCRIPTION
**Pull request by yovanoc**
This PR fix #2445.

I don't know if it's good but the dist/preview release folder was deleted, but I tested my ParticleHelper on a separate project after building it with "gulp typescript" and all works great!

In the config.json I am not really sure about the dependUpon things

I haven't done the json part of the config but I will do that soon :)

Here's my first try : https://github.com/yovanoc/Babylon.js/blob/dc4e33cd7f5e9bbddc6cf0966717fdbb89af5022/src/Helpers/babylon.particleHelper.ts

### CODE REVIEW

**Comment by deltakosh** [comment]
We cannot work on this PR as it delete files :). Can you try to build a new one? There are too many files changed

**Comment by yovanoc** [comment]
I understand but I try gulp and gulp typescript what command build all this folder ?

**Comment by deltakosh** [comment]
This seems more related to a promise blemishes with your fork. The dis/preview should not have been deleted

**Comment by yovanoc** [comment]
Better now?

**Comment by deltakosh** [comment]
Can you completely exclude all the dist folder from this PR?

**Comment by yovanoc** [comment]
Seems better now I think

File: src/Helpers/babylon.particleHelper.ts [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+        /**
+         * This is the main static method (one-liner) of this helper to create different particle systems.
+         * @param type This string will be parsed to a ParticleSystemType
+         * @param emitter The object where the particle system will start to emit from.
+         * @param scene The scene where the particle system should live.
+         * @param gpu If the system will use gpu.
+         * @returns the ParticleSystem created.
+         */
+        public static Create(type: string, emitter: AbstractMesh, scene: Nullable<Scene> = Engine.LastCreatedScene, gpu: boolean = false): ParticleSystem {
```
Comment (deltakosh) [inline src/Helpers/babylon.particleHelper.ts]: I would rather prefer having a json file loaded where you can find fire or smoke parameters

Does it make sense?

Comment (deltakosh) [reply src/Helpers/babylon.particleHelper.ts]: Create will be async and you can just call Tools.LoadFile to get the content of the json file (from 
baseAssetsUrl + "/systems.json" for instance)
You can rely on promise to get a CreateAsync function for isntance

Comment (yovanoc) [reply src/Helpers/babylon.particleHelper.ts]: Oh okay, i haven't see that, thanks you :)

Comment (yovanoc) [reply src/Helpers/babylon.particleHelper.ts]: But, can I know why you prefer async method + json file over a simple synchronous method?

**Comment by deltakosh** [comment]
This helps to not bloat the engine core code.
We can add 100 different configuration and the file will just be loaded when needed

**Comment by yovanoc** [comment]
You're right, thanks. And for the unit tests?

**Comment by yovanoc** [comment]
I have to "gulp typescript" to generate what I've done in order to unit tests ... So after my tests, I have to discard all of the generated dist/preview release for my PR? :o

**Comment by deltakosh** [comment]
You have to discard it before PR

**Comment by yovanoc** [comment]
Yes, but why can't I see a console.log in a unit test to see something? Is the command "gulp tests-unit"?

**Comment by deltakosh** [comment]
It is the right order and it will play all tests and return messages only if there is a problem

**Comment by yovanoc** [comment]
I added the line of my test file in the karma conf and replace the .ts by .js like others files and after I have this error when I launch the tests :
 
11 05 2018 15:35:08.862:WARN [watcher]: Pattern "/home/devchris/Desktop/Babylon.js/tests/unit/babylon/src/Helpers/babylon.particleHelper.tests.js" does not match any file.

**Comment by RaananW** [comment]
Where is your ts file located?
Can you also try running `gulp tests-unit-transpile` and see if any error is thrown?

**Comment by yovanoc** [comment]
```bash
devchris@ubuntu:~/Desktop/Babylon.js/Tools/Gulp$ gulp tests-unit-transpile
[16:49:53] Using gulpfile ~/Desktop/Babylon.js/Tools/Gulp/gulpfile.js
[16:49:53] Starting 'tests-unit-transpile'...
[16:49:56] Finished 'tests-unit-transpile' after 3.03 s
devchris@ubuntu:~/Desktop/Babylon.js/Tools/Gulp$ 
```

All good for that, and my file is at : 

/home/devchris/Desktop/Babylon.js/tests/unit/babylon/src/Helpers/babylon.particleHelper.tests.ts

**Comment by yovanoc** [comment]
Oh, I delete and recreate file and it's work :thinking:

**Comment by RaananW** [comment]
I don't see the file commited in this PR. 

Maybe you could commit the assets in a different PR and then you will have the assets online

**Comment by yovanoc** [comment]
Yeah but I have it in local

File: Tools/Gulp/config.json [hunk 2]
```diff
@@ -1170,6 +1171,16 @@
                 "additionalTextures"
             ]
         },
+        "particleHelper": {
+            "files": [
+                "../../src/Helpers/babylon.particleHelper.js"
+            ],
+            "dependUpon": [
+                "core",
+                "backgroundMaterial",
```
Comment (deltakosh) [inline Tools/Gulp/config.json]: you should just rely on particles actually

File: dist/preview release/what's new.md [hunk 3]
```diff
@@ -14,6 +14,7 @@
 - Added predicate function to customize the list of mesh included in the computation of bounding vectors in the ```getHierarchyBoundingVectors``` method ([sebavan](http://www.github.com/sebavan))
 - Added webVR constructor options: disable laser pointer toggle, teleportation floor meshes ([TrevorDev](https://github.com/TrevorDev))
 
+
```
Comment (deltakosh) [inline dist/preview release/what's new.md]: Please add a line in the major update to present your great work?

File: src/Helpers/babylon.particleHelper.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
+module BABYLON {
+    /**
+     * Represents all the data needed to create a ParticleSystem.
+     */
+    export interface IParticleSystemData {
+        type: ParticleSystemType;
```
Comment (deltakosh) [inline src/Helpers/babylon.particleHelper.ts]: We should also be able to define the emitter type (+ parameters)

File: src/Helpers/babylon.particleHelper.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        minAngularSpeed: number;
+        maxAngularSpeed: number;
+        minEmitPower: number;
+        maxEmitPower: number;
+        updateSpeed: number;
+    }
+    /**
+     * ParticleSystemType
+     */
+    export enum ParticleSystemType {
```
Comment (deltakosh) [inline src/Helpers/babylon.particleHelper.ts]: Not needed anymore

File: src/Helpers/babylon.particleHelper.ts [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+        /**
+         * This is the main static method (one-liner) of this helper to create different particle systems.
+         * @param type This string will be parsed to a ParticleSystemType
+         * @param emitter The object where the particle system will start to emit from.
+         * @param scene The scene where the particle system should live.
+         * @param gpu If the system will use gpu.
+         * @returns the ParticleSystem created.
+         */
+        public static Create(type: string, emitter: AbstractMesh,
```
Comment (deltakosh) [inline src/Helpers/babylon.particleHelper.ts]: CreateAsync

File: src/Helpers/babylon.particleHelper.ts [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+                Tools.LoadFile(`${this._baseAssetsUrl}/systems/${typeParsed}.json`, (data, response) => {
+                    const newData = JSON.parse(data.toString()) as IParticleSystemData;
+                    return resolve(this._createSystem(newData));
+                });
+
+            });
+        }
+
+        private static _parseType(type: string): ParticleSystemType {
```
Comment (deltakosh) [inline src/Helpers/babylon.particleHelper.ts]: not needed

File: src/Helpers/babylon.particleHelper.ts [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+                case "SMOKE":
+                    return ParticleSystemType.Smoke;
+                default:
+                    return ParticleSystemType.None;
+            }
+        }
+
+        private static _createSystem(data: IParticleSystemData): ParticleSystem {
+            // Create a particle system
+            const fireSystem = new ParticleSystem(data.type, data.capacity, this._scene);
```
Comment (deltakosh) [inline src/Helpers/babylon.particleHelper.ts]: rename to system

File: Tools/Gulp/config.json (hunk 2 above)
Comment (yovanoc) [reply Tools/Gulp/config.json]: yeah that's what I said I was not sure at all what to do here so just : 
```json
"dependUpon": [
    "particles"
]
```

is good?

File: dist/preview release/what's new.md (hunk 3 above)
Comment (yovanoc) [reply dist/preview release/what's new.md]: Oh I will do that :)

File: src/Helpers/babylon.particleHelper.ts (hunk 4 above)
Comment (yovanoc) [reply src/Helpers/babylon.particleHelper.ts]: I'm not sure how to do this?
Can you give me an example please?

File: src/Helpers/babylon.particleHelper.ts (hunk 6 above)
Comment (yovanoc) [reply src/Helpers/babylon.particleHelper.ts]: Oh yeah sorry

File: src/Helpers/babylon.particleHelper.ts (hunk 5 above)
Comment (yovanoc) [reply src/Helpers/babylon.particleHelper.ts]: Mhhh I was thinking this was a good idea to let this with the parse method but yeah if a user made a typo is not our fault ^^

File: src/Helpers/babylon.particleHelper.ts (hunk 8 above)
Comment (yovanoc) [reply src/Helpers/babylon.particleHelper.ts]: of course!

File: src/Helpers/babylon.particleHelper.ts [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+         * @returns the ParticleSystem created.
+         */
+        public static CreateAsync(type: string, emitter: AbstractMesh,
+                                   scene: Nullable<Scene> = Engine.LastCreatedScene, gpu: boolean = false): Promise<ParticleSystem> {
+            
+            return new Promise((resolve, reject) => {
+                if (scene) {
+                    this._scene = scene;
+                } else {
+                    throw new Error("A particle system need a scene.");
```
Comment (deltakosh) [inline src/Helpers/babylon.particleHelper.ts]: Should be reject() here

File: dist/preview release/what's new.md [hunk 10]
```diff
@@ -13,6 +13,7 @@
 - Added ability to not generate polynomials harmonics upon prefiltered texture creation ([sebavan](http://www.github.com/sebavan))
 - Added predicate function to customize the list of mesh included in the computation of bounding vectors in the ```getHierarchyBoundingVectors``` method ([sebavan](http://www.github.com/sebavan))
 - Added webVR constructor options: disable laser pointer toggle, teleportation floor meshes ([TrevorDev](https://github.com/TrevorDev))
+- Added a ParticleHelper class to create some pre-configured particle systems in a one-liner method style ([DevChris](https://github.com/yovanoc))
```
Comment (deltakosh) [inline dist/preview release/what's new.md]: You can promote it to Major updates :)

File: src/Helpers/babylon.particleHelper.ts (hunk 4 above)
Comment (deltakosh) [reply src/Helpers/babylon.particleHelper.ts]: It is not the emitter but the emitter shape. The shape that the particles will use

**Comment by yovanoc** [comment]
I will have to do GPU after, but if I replace the return type ParticleSystem by the IParticleSystem interface its telling me that some properties is missing between GPU and normal systems

**Comment by deltakosh** [comment]
Can you tell more about the missing properties?

**Comment by yovanoc** [comment]
minEmitBox, maxEmitBox, direction1, direction2, minAngularSpeed, maxAngularSpeed

and all the create...Emitter methods that you tell me to do now

**Comment by deltakosh** [comment]
Ok let's first see the CPU part..we will deal with GPU afterwards

**Comment by yovanoc** [comment]
Yes, you're right, I will push soon :)

**Comment by yovanoc** [comment]
But now all emitters variables in the switch are not used and Travis doesn't like that x)

File: src/Helpers/babylon.particleHelper.ts [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+            // Angular speed, in radians
+            system.minAngularSpeed = data.minAngularSpeed;
+            system.maxAngularSpeed = data.maxAngularSpeed;
+
+            // Speed
+            system.minEmitPower = data.minEmitPower;
+            system.maxEmitPower = data.maxEmitPower;
+            system.updateSpeed = data.updateSpeed;
+
+            switch (data.emitterType) {
```
Comment (deltakosh) [inline src/Helpers/babylon.particleHelper.ts]: no need for an enum here. Plain string is fine

Comment (yovanoc) [reply src/Helpers/babylon.particleHelper.ts]: good like that?

Comment (deltakosh) [reply src/Helpers/babylon.particleHelper.ts]: yes!

Comment (yovanoc) [reply src/Helpers/babylon.particleHelper.ts]: ahah great

File: tests/unit/babylon/src/Helpers/babylon.particleHelper.tests.ts [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+          textureSize: 256,
+          deterministicLockstep: false,
+          lockstepMaxSteps: 1
+      });
+  });
+
+  describe('#JSON', () => {
+      it('creates a fire particle system', () => {
+          const scene = new BABYLON.Scene(subject);
+          //
```
Comment (deltakosh) [inline tests/unit/babylon/src/Helpers/babylon.particleHelper.tests.ts]: need to be completed right?

Comment (yovanoc) [reply tests/unit/babylon/src/Helpers/babylon.particleHelper.tests.ts]: Yes I will do that when assets will be online

**Comment by deltakosh** [comment]
We will also need detailed doc when everything will be ready

**Comment by yovanoc** [comment]
You are right!  I'll do unit tests for now so! :)

**Comment by deltakosh** [comment]
Ping:) Are you working on the documentation?

**Comment by yovanoc** [comment]
Yeah I'm a little bit busy now, I've some work to do, but I will do it :)

**Comment by yovanoc** [comment]
I've a lot of work but tell me where I have to add the documentation and I will do it :)

**Comment by deltakosh** [comment]
Probably here: https://github.com/BabylonJS/Documentation/blob/master/content/babylon101/babylon101/Particles.md

**Pull request merged**